src/ngx_http_lua_kong_log.c

```diff
@@ -14,45 +14,122 @@
+ngx_http_lua_kong_ffi_set_dynamic_log_level(int log_level, int timeout)
 {
-    ngx_log_t                   *log;
-    ngx_listening_t             *ls;
-    ngx_http_core_loc_conf_t    *clcf;
-    ngx_uint_t                   i;
+    ngx_http_lua_kong_main_conf_t *lkmcf =
+        ngx_http_cycle_get_module_main_conf(ngx_cycle, ngx_http_lua_kong_module);
 
-    if (r == NULL) {
-        return NGX_ERROR;
-    }
+    ngx_http_lua_kong_assert(lkmcf != NULL);
 
-    clcf = ngx_http_get_module_loc_conf(r, ngx_http_core_module);
-    clcf->error_log->log_level = (ngx_uint_t) level;
+    if (timeout == 0) {
+        /*
+        ** if the timeout is 0,
+        ** we will disable the dynamic log level,
+        ** that is, the log_level will be restored to the default setting from log_level in kong.conf.
+        ** and ignore the first parameter.
+        */
+        lkmcf->dynamic_log_level_set = 0;
+        return NGX_OK;
+    }
 
-    if (ngx_cycle == NULL) {
+    /*
+    ** if the log level is invalid,
+    ** we will not change the current log level,
+    ** and return an error.
+    */
+    if (log_level > NGX_LOG_DEBUG || log_level < NGX_LOG_STDERR) {
         return NGX_ERROR;
     }
 
-    if (r->connection && r->connection->log) {
-        log = r->connection->log;
+    lkmcf->dynamic_log_level = log_level;
+    lkmcf->dynamic_log_level_timeout_at = (time_t)ngx_time() + (time_t)timeout;
+    lkmcf->dynamic_log_level_set = 1;
+
+    return NGX_OK;
+}
+
+ngx_uint_t
+ngx_http_lua_kong_get_dynamic_log_level(ngx_uint_t current_log_level)
+{
+    if (!g_ready_for_dynamic_log_level) {
+        /* 
+        ** This function might be called before the
+        ** neccessary resources are ready,
+        ** such as the configuration,
+        ** the data structure for HTTP module, etc.
+         */
+        return current_log_level;
+    }
 
-    } else {
-        log = ngx_cycle->log;
+    ngx_http_lua_kong_main_conf_t *lkmcf =
+        ngx_http_cycle_get_module_main_conf(ngx_cycle, ngx_http_lua_kong_module);
+    
+    if (lkmcf == NULL) {
+        /*
+        ** This function might be called outside the HTTP context.
+        */
+        return current_log_level;
     }
 
-    /* current request */
-    log->log_level = (ngx_uint_t) level;
+    if (!lkmcf->dynamic_log_level_set) {
+        return current_log_level;
+    }
 
-    /* for each listening socket (new requests) */
-    ls = ngx_cycle->listening.elts;
-    for (i = 0; i < ngx_cycle->listening.nelts; i++) {
-        ls[i].log.log_level = (ngx_uint_t) level;
-        ls[i].logp->log_level = (ngx_uint_t) level;
+    if (lkmcf->dynamic_log_level_timeout_at < ngx_time()) {
+        lkmcf->dynamic_log_level_set = 0;
+        return current_log_level;
     }
 
-    return NGX_OK;
+    return lkmcf->dynamic_log_level;
+}
+
+int
+ngx_http_lua_kong_ffi_get_dynamic_log_level(int current_log_level)
+{
+    return (int)ngx_http_lua_kong_get_dynamic_log_level((ngx_uint_t)current_log_level);
```

Comment:
```suggestion
    return (int)ngx_http_lua_kong_get_dynamic_log_level(current_log_level);
```

I think the cast is unnecessary, see lua-ngx-module `ngx_http_lua_ffi_errlog_set_filter_level`